### PR TITLE
Add pseudospectral method for trajectory optimization

### PR DESCRIPTION
**Pull request by shensquared**
- See example/glider/runPSM for an example call
- Todo: Add more detailed comments for key functions
- Todo: Fix xtraj and utraj display reconstruction
- Todo: consolidate files

### CODE REVIEW

**Comment by RussTedrake** [comment]
the code looks great.  
how exactly were the methods from Herber obtained?  I think perhaps we should put them in the thirdParty folder instead?

**Comment by shensquared** [comment]
Thanks!

The methods were downloaded directly from Matlab file exchange center, and then tested against Wolfram-alpha for their accuracy. Herber didn't write the methods I used though, the author's header were batch generated for his entire project. The book authors he cited under the functions were the origianl authors, so I've moved those comments out front for more prominent recognition. 

I'll put those methods in the thirdPaty folder and open a new pull request (will also add more clear comments for PSMTraj method for readabitly)

File: drake/solvers/trajectoryOptimization/PsmTraj.m [hunk 1]
```diff
@@ -0,0 +1,445 @@
+classdef PsmTraj < NonlinearProgram
+  % Pseudospectral method
```
Comment (hongkai-dai) [inline drake/solvers/trajectoryOptimization/PsmTraj.m]: Why `PsmTraj` is not a subclass of `DirectTrajectoryOptimization`? It fits the `DirectTrajectoryOptimization` model, and shares a lot of the functions. Also consider to rename it to `PseudoSpectralMethodTrajOpt`. `PSM` is not a common abbreviation.

File: drake/solvers/trajectoryOptimization/PsmTraj.m [hunk 2]
```diff
@@ -0,0 +1,445 @@
...
+        end       
+      end 
+      
+    %%difference between the dynamic and collocation
+      g=[xdot_p]-[xdot_f];
+    %% gradient of such difference
+      dxandudot_f=[dxdh_f,dxdot_f,udot_f];
+      dD_p=[zeros(N*nX,1),dDxx_p,dDxu_p];
+      dg=sparse(dD_p)-sparse(dxandudot_f);      
+    end
```
Comment (hongkai-dai) [inline drake/solvers/trajectoryOptimization/PsmTraj.m]: You do not need a `sparse` function call here. The sparsity should be specified in `cnstr = cnstr.setSparseStructure(rows,cols)`, after you defined `cnstr`  in line 69.

File: drake/solvers/trajectoryOptimization/PsmTraj.m [hunk 3]
```diff
@@ -0,0 +1,445 @@
...
+      dDxx_p=sparse(N*nX,N*nX);
+      dDxu_p=sparse(N*nX,N*nU);
+
+    %%f and df (the dynamics) evaluated at the knot points  
+      for k=1:N, %row indices
+        xdot_f((k-1)*nX+1:k*nX,:)= varargin{k}.xdot*scale; %%f(x,u), with scaling
+        dxdh_f((k-1)*nX+1:k*nX,:)= varargin{k}.xdot/initial_h1; %%df/dh 
+        dxdot_f((k-1)*nX+1:k*nX,(k-1)*nX+1:k*nX) = varargin{k}.dxdot(:,2:1+nX)*scale;%%df/dx, with scaling
+        udot_f((k-1)*nX+1:k*nX,(k-1)*nU+1:k*nU) = varargin{k}.dxdot(:,2+nX:end)*scale; %%df/du 
+        for j=1:N, %column indices
```
Comment (hongkai-dai) [inline drake/solvers/trajectoryOptimization/PsmTraj.m]: vectorize this for loop?

File: drake/solvers/trajectoryOptimization/lepoly.m [hunk 4]
```diff
@@ -0,0 +1,39 @@
+function [varargout]=lepoly(n,x)
```
Comment (hongkai-dai) [inline drake/solvers/trajectoryOptimization/lepoly.m]: Fix the indentation of this file.

File: drake/solvers/trajectoryOptimization/CGL_nodes.m [hunk 5]
```diff
@@ -0,0 +1,30 @@
...
+% Author: Daniel R. Herber, Graduate Student, University of Illinois at
+% Urbana-Champaign
+% Date: 06/04/2015
+%-------------------------------------------------------------------------
+function tau = CGL_nodes(obj)
+
+	N=obj.N-1;
+    syms n
+    % calculate node locations
+    for k = 0:N
```
Comment (hongkai-dai) [inline drake/solvers/trajectoryOptimization/CGL_nodes.m]: vectorize this loop

File: drake/solvers/trajectoryOptimization/CGL_nodes.m [hunk 6]
```diff
@@ -0,0 +1,30 @@
...
+% tau = CGL_nodes(2)
+% -1     0     1
+% tau = CGL_nodes(3)
+% -1  -0.5   0.5   1
+%--------------------------------------------------------------------------
+% Author: Daniel R. Herber, Graduate Student, University of Illinois at
+% Urbana-Champaign
+% Date: 06/04/2015
+%-------------------------------------------------------------------------
+function tau = CGL_nodes(obj)
```
Comment (hongkai-dai) [inline drake/solvers/trajectoryOptimization/CGL_nodes.m]: This function uses symbolic computation first, and then substitute the numerical values. Consider to rewrite it with numeric computation directly.

File: drake/solvers/trajectoryOptimization/PsmTraj.m [hunk 7]
```diff
@@ -0,0 +1,445 @@
+classdef PsmTraj < NonlinearProgram
+  % Pseudospectral method
+  % Knot pointes are determined by N-th order orthogonal polynomials, e.g. Legendre polynomials, Chebyshev polynomials
+  % Then at each knot points, use Lagrange polynomials to approximate state values. 
+  % f(x(k),u(k)) and f(x(k+1),u(k+1)) are evaluated at the knot points to match the derivative of the Lagrange poly
+   %%options,'ps_method' detemines if to use Legendre (option.ps_method=1) or to use Chebyshev (option.ps_method=2)
+
+  properties (SetAccess = protected)
+    N       % number of timesteps
+    options % options, yup
```
Comment (hongkai-dai) [inline drake/solvers/trajectoryOptimization/PsmTraj.m]: Add documentation for options

File: drake/solvers/trajectoryOptimization/PsmTraj.m [hunk 8]
```diff
@@ -0,0 +1,445 @@
...
+  end
+  
+  methods
+    function obj = PsmTraj(plant,N,durations,options)
+      if isscalar(durations), durations=[durations,durations]; end
+      if nargin < 4
+        options = struct();
+      end
+      if ~isfield(options,'ps_method')
+        options.ps_method = 1;
```
Comment (hongkai-dai) [inline drake/solvers/trajectoryOptimization/PsmTraj.m]: Use string 'chebyshev' or 'legendre' instead of value '1' or '2'. Or you may derive two classes `ChebyshevPseudoSpectralTraj` and `LegendrePseudoSpctralTraj` from `PsmTraj`

**Comment by hongkai-dai** [comment]
Code is great. The test `runPSM` is on Legendre polynomial, consider to add a test on Chebyshev polynomial.

File: drake/solvers/trajectoryOptimization/PsmTraj.m (hunk 1 above)
Comment (shensquared) [inline drake/solvers/trajectoryOptimization/PsmTraj.m]: Thanks for the suggestions, Hongkai. Will rename the class. I also agree with you on the proper hierarchy, I initially did attempt it as a `DirectTrajectoryOptimization` subclass, then ran into a few issues I didn't know a quick fix and rushed into this ugly hack. Will fix it now that there's less time pressure.

File: drake/solvers/trajectoryOptimization/PsmTraj.m (hunk 2 above)
Comment (shensquared) [inline drake/solvers/trajectoryOptimization/PsmTraj.m]: Cool! I did not know that `cnstr.setSparseStructure` trick. I'll definitely try that. 

I'll also try all the suggestions below, thanks for all of them!

**Comment by shensquared** [comment]
I'll try the wonderful suggestions Hongkai offered and update after the fix.  

I also just realized I should add the thirdParty code license, will fix that too. 

It also seems (based on some online discussions) that I did not need to track down and promote the original authors whose work were folded in the redistribution? It's only required to credit the direct source?

**Comment by RussTedrake** [comment]
correct.  we have other items from the matlab exchange in the thirdParty folder. as long as you keep the license file along with the source, it should be good.
also - you do not need to open a new pull request.  you can simply push new commits to the branch where you made this request and it will update everything while keeping the discussion thread intact.

**Comment by shensquared** [comment]
Fixed/optimized the items @hongkai-dai suggested, and elaborated on the code comments for better readability. Two major things I noticed after the project report:

- Hongkai's suggestion of vectorization and setting constraint gradient sparse structre sped up the NLP solver significantly, in some cases by over 40%. But it is still not as fast as DirCol in general.

- Despite interpolating at Gauss quadratures, reconstructing the trajectory by Lagrange interpolation may still suffer from Runge's phenomenon when the number of knot points is big. An example of such 'fail' case is commented in the example call `runPSM` in /drake/examples/glider folder in case the user isn't aware of the risk.

In summary, with further investigation, I still prefer DirCol for its generally faster and more reliable performance.

File: drake/solvers/trajectoryOptimization/PseudoSpectralMethodTrajOpt.m [hunk 9]
```diff
@@ -0,0 +1,221 @@
...
+  methods
+    function obj = PseudoSpectralMethodTrajOpt(plant,N,durations,options)
+
+      if nargin < 4
+        options = struct();
+      end
+      if ~isfield(options,'ps_method')
+        options.ps_method = PseudoSpectralMethodTrajOpt.LGL;
+      end
+      if ~isfield(options,'time_option')| options.time_option ~= 3
```
Comment (hongkai-dai) [inline drake/solvers/trajectoryOptimization/PseudoSpectralMethodTrajOpt.m]: Add a check here, throw an error if options.ps_method is not LGL or CGL

File: drake/solvers/trajectoryOptimization/PseudoSpectralMethodTrajOpt.m [hunk 10]
```diff
@@ -0,0 +1,221 @@
...
+      S_time=[ratio,zeros(N-1,N-2)]-eye(N-1); % so that the time steps will keep the inter-point ratio as the initial one obtained from Gauss quadatures
+      scale_constaint = LinearConstraint([durations(1);-inf; zeros(N-2,1)],[durations(2);inf;zeros(N-2,1)], [A_time;S_time]);
+      obj = obj.addConstraint(scale_constaint,obj.h_inds);  
+    end
+
+    function obj = addDynamicConstraints(obj)
+      N = obj.N;
+      nX = obj.plant.getNumStates();
+      nU = obj.plant.getNumInputs();
+      options=obj.options;
```
Comment (hongkai-dai) [inline drake/solvers/trajectoryOptimization/PseudoSpectralMethodTrajOpt.m]: There is overhead to copy the options structure.

File: drake/solvers/trajectoryOptimization/PseudoSpectralMethodTrajOpt.m [hunk 11]
```diff
@@ -0,0 +1,221 @@
...
+      utraj = utraj.setOutputFrame(obj.plant.getInputFrame);
+    end
+    
+    function xtraj = reconstructStateTrajectory(obj,z)
+      % Interpolate all knot points to reconstruct a trajectory using
+      % Lagrange interpolation
+      nX = obj.plant.getNumStates();
+      x = reshape(z(obj.x_inds),[],obj.N);
+      % similar reason as in reconstructInputTrajectory() for the choice of interpolating times
+      xtraj=PPTrajectory(LagrangeInterpolation(obj.tau,x,nX));
```
Comment (hongkai-dai) [inline drake/solvers/trajectoryOptimization/PseudoSpectralMethodTrajOpt.m]: Does this reconstruct Legendre polynomial only? How about Chebyshev polynomials?

File: drake/util/LagrangeInterpolation.m [hunk 12]
```diff
@@ -0,0 +1,29 @@
...
+else
+  if (prod(dim)~=n) error('dim does not match size of y'); end
+end
+if (length(size(y))~=2) 
+  y = reshape(y,n,[]);
+end
+if (size(y,2)~=length(t)) error('last dimension of y should be the same size as t'); end
+
+
+for i=1:(dim)
```
Comment (hongkai-dai) [inline drake/util/LagrangeInterpolation.m]: I believe you can vectorize this for loop also. I wrote a function to compute the Lagrange polynomial before, you could refer to it in http://www.mathworks.com/matlabcentral/fileexchange/36522-compute-the-value-of-lagrange-polynomial-at-given-points/content/lagrangePolyVal.m

**Comment by hongkai-dai** [comment]
Code looks great!

I also observed the Runge's phenomena at the ends of the trajectory, when I worked on pseudo spectral method before.

**Comment by RussTedrake** [comment]
Thank you both.  @shensquared -- will wait for the small mods from @hongkai-dai 's code review, then merge.

File: drake/util/LagrangeInterpolation.m (hunk 12 above)
Comment (hongkai-dai) [inline drake/util/LagrangeInterpolation.m]: On a second thought, vectorizing this code might be not necessary, as it is called for only once in each trajectory optimization.

**Comment by shensquared** [comment]
Thanks for the suggestions. The code is now all updated.
